src/inference/tempering.jl

```diff
@@ -0,0 +1,166 @@
+###########################
+#   SAMPLING
+###########################
+
+"""
+Upon calling sample on a `TemperedSampler`, we must manually instantiate a `Sampler`
+to insert into the `TemperedSampler`, rather than wrapping the `TemperedSampler`
+itself in a `Sampler`
+"""
+function create_tempered_sampler(
+    model::AbstractModel,
+    sampler::MCMCTempering.TemperedSampler{<:InferenceAlgorithm}
+)
+    return MCMCTempering.TemperedSampler(
+        Sampler(sampler.internal_sampler, model),
+        sampler.Δ,
+        sampler.Δ_init,
+        sampler.N_swap,
+        sampler.swap_strategy
+    )
+end
+
+
+function AbstractMCMC.sample(
+    rng::AbstractRNG,
+    model::AbstractModel,
+    sampler::MCMCTempering.TemperedSampler{<:InferenceAlgorithm},
+    N::Integer;
+    chain_type=MCMCChains.Chains,
+    resume_from=nothing,
+    progress=PROGRESS[],
+    kwargs...
+)
+    tempered_sampler = create_tempered_sampler(model, sampler)
+    if resume_from === nothing
+        return AbstractMCMC.sample(rng, model, tempered_sampler, N;
+                                   chain_type=chain_type, progress=progress, kwargs...)
+    else
+        return resume(resume_from, N; chain_type=chain_type, progress=progress, kwargs...)
+    end
+end
+
+
+function AbstractMCMC.sample(
+    rng::AbstractRNG,
+    model::AbstractModel,
+    sampler::MCMCTempering.TemperedSampler{<:InferenceAlgorithm},
+    ensemble::AbstractMCMC.AbstractMCMCEnsemble,
+    N::Integer,
+    n_chains::Integer;
+    chain_type=MCMCChains.Chains,
+    progress=PROGRESS[],
+    kwargs...
+)
+    tempered_sampler = create_tempered_sampler(model, sampler)
+    return AbstractMCMC.sample(rng, model, tempered_sampler, ensemble, N, n_chains;
+                               chain_type=chain_type, progress=progress, kwargs...)
+end
+
+
+
+###########################
+#   MODEL
+###########################
+
+struct TemperedEval{T<:Real}
+    model :: DynamicPPL.Model
```

Comment:
This is part of my suggestions:
```suggestion
struct TemperedEval{T<:Real,F}
    f::F
```